### PR TITLE
feat: add storybook framework

### PR DESCRIPTION
**Pull request by JSerFeng**
## Summary
Add support for storybookV7 framework

storybook
- framework: used for storybook framework
- builder: used for storybook builder

### <samp>🤖 Generated by Copilot at a3d670a</samp>

This pull request adds a new `storybook-builder` package that integrates modern.js with storybook, using either webpack or rspack as the bundler. It also updates some existing packages to support this integration, such as exporting types and functions, adding node exports, and using types from the same package. Additionally, it adds a changeset file, an eslint configuration file, and a custom docgen handler and loader for storybook.

### <samp>🤖 Generated by Copilot at a3d670a</samp>

*  Add a changeset file to describe the pull request ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-f217050781ecdbd5c2dbe6753ffdb45655da3989b81c3d8386c84a4bade63d05R1-R6))
*  Export loadEnv and loadConfig functions from the core package to load the modern.js configuration and environment variables ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-bae066e7ed5bbce786c45a5294f6edbc219e775b87f1883eff1a5ac43806ebe5L4-R6), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-412d60d03940b7df6ed82f40c651475eac6656124e356884c19c8f227435f668R4))
*  Add node exports for various entry points of the plugin-runtime package to support node.js environments ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R30), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R40), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R46), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R52), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561L72-R76), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R83), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561L84-R95), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561R101), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a90f331e2d4e9320efe81c8af64349df4b94c4c63c247993e91f00bcfc1a8561L99-R108))
*  Define the storybook-builder package with a package.json file and an eslint configuration file ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-1d54535d7b406252bee2755120c26e35eef63f97c7ccab7a6b3538bc9213beacR1-R94), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-97d579d4382bb650b4ae5b3541e2916a574c355ca094bf6f030b16e6be1a53eeR1-R7))
*  Implement the build and start functions for the storybook builder using webpack or rspack as the bundler ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-872331c879ec9c563376a7288d6cc7e8717273ed660034e5f9fb32021ee94e03R1-R114))
*  Implement the getCompiler function that creates and returns a webpack or rspack compiler for storybook using the modern.js builder and plugins ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-d14ccbdf163a4c757552d040dfa1e8c1141edff82490a6dc18a0b95d744c7ff2R1-R50))
*  Implement the applyDocgenWebpack and applyDocgenRspack functions that apply the react-docgen or react-docgen-typescript plugins to the bundler configuration for storybook ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-feaf3c8e96b9a9072df2321787ef2a783f453fc7e7885c56b16bfe5b6b566c59R1-R108))
*  Implement a custom loader that runs the react-docgen plugin on the source code using a tinypool process and appends the generated docs to the source code ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-eb7e6e1deefb1c97786227ec92d4f85bb5d191aad6954c6849097b9ce844c58dR1-R25))
*  Implement a custom handler for react-docgen that sets the actualName property on the generated docs based on the component's actual name ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-a9a3ada8ee0e628a1bc83d4276f923a01216f0bda76c97083bf3c84f6f847b6bR1-R57))
*  Implement a decorator function that wraps the storybook components with modern.js runtime providers using the WrapProviders component ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-708ae807c0832bfce5f8958dc64cda7aff00078f18a205730d07468baf5a8d3eR1-R20))
*  Implement a modern.tsx file that wraps the storybook components with modern.js runtime providers ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-7203caeb64f3e730a7075de0b1353926992be880ae873d101b635856654bf4abR1-R56))
*  Define the decorators for the modern.js runtime addon in a preview.ts file ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-805ba659fb2bd1cc6a1a1eb40eb73f335f56e1e153349ab212f5613b21815589R1-R6))
*  Define the addon ID for the modern.js runtime addon in a constants.ts file ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-b4c19bb397aad76bdd62e405e75a9bfeeff6ae16bf9af16bcbbc2a6f26103981R1))
*  Export an empty object from the addons index.ts file to avoid isolatedModules error ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-ba86d67862a0601a3a810d61cd4d72e345ec6851e3dda6f9956ab9e929422bfdR1-R8))
*  Define the interface for the modern.js configuration options for storybook in a type.ts file ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-f445c8508fc53d32287c8e2fe250898207e77540e1d985e7166beba4b3200b22R1-R4))
*  Import and export various types from the builder-rspack-provider and builder-webpack-provider packages to use and create the rspack and webpack compilers for storybook ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-639e063e992304dae8479eae5392cb49de7c290f5901af842eec5f7e25fe2f4cR12), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-639e063e992304dae8479eae5392cb49de7c290f5901af842eec5f7e25fe2f4cR19-R21), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-753e5e69851353e5879f3c314ec36e4b9df1b5848a2ff1cd087950d9583801b9L3-R7), [link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-753e5e69851353e5879f3c314ec36e4b9df1b5848a2ff1cd087950d9583801b9L22-R32))
*  Remove an empty line from the app-tools package ([link](https://github.com/web-infra-dev/modern.js/pull/4585/files?diff=unified&w=0#diff-797ceb5288bf4f83c85aba37cbeae7f82928a9e06542582debfcf58b75b817f8L26))

## Checklist

- [x] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Comment by chenjiahan** [comment]
There is a warning when use `bundler: 'rspack'`:

image

**Comment by chenjiahan** [comment]
Can we remove the assets log of storybook? It is quite annoying.

image

File: packages/builder/builder-webpack-provider/src/types/thirdParty/index.ts [hunk 1]
```diff
@@ -1,6 +1,10 @@
 import type WebpackChain from '@modern-js/utils/webpack-chain';
 import type webpack from 'webpack';
-import type { Configuration as WebpackConfig } from 'webpack';
+import type {
+  Configuration as WebpackConfig,
+  Compiler as WebpackCompiler,
+  Compiler as WebpackMultiCompiler,
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/types/thirdParty/index.ts:6]: Why not `MultiCompiler as WebpackMultiCompiler`

File: packages/storybook/builder/package.json [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      "jsnext:source": "./src/types.ts",
+      "types": "./dist/types.d.ts",
+      "default": "./dist/types.js"
+    },
+    "./templates/preview.ejs": "./templates/preview.ejs",
+    "./templates/virtualModuleModernEntry.js.handlebars": "./templates/virtualModuleModernEntry.js.handlebars",
+    "./package.json": "./package.json"
+  },
+  "engines": {
+    "node": ">=16.0.0"
```
Comment (chenjiahan) [inline packages/storybook/builder/package.json:35]: Is Node 14 no longer supported by storybook v7?

File: packages/storybook/builder/src/plugin-storybook.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+      providerImportSource: '@storybook/addon-docs/mdx-react-shim',
+      ...mdxPluginOptions?.mdxCompileOptions,
+      remarkPlugins: [
+        remarkSlug,
+        remarkExternalLinks,
+        ...(mdxPluginOptions?.mdxCompileOptions?.remarkPlugins ?? []),
+      ],
+    },
+  });
+  const mdxLoader = options.features?.legacyMdx1
```
Comment (chenjiahan) [inline packages/storybook/builder/src/plugin-storybook.ts:246]: Do we need to compat legacyMdx1?

File: tests/integration/storybook/package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  "devDependencies": {
+    "@modern-js/storybook": "workspace:*",
+    "@modern-js/builder-plugin-swc": "workspace:*",
+    "@storybook/addon-essentials": "^7.4.0",
+    "@storybook/addon-interactions": "^7.4.0",
+    "@storybook/addon-links": "^7.4.0",
+    "@storybook/blocks": "^7.4.0",
+    "@storybook/jest": "^0.1.0",
+    "@storybook/react": "^7.4.0",
+    "@storybook/testing-library": "^0.2.0",
```
Comment (chenjiahan) [inline tests/integration/storybook/package.json:33]: Why user need to install so many @storybook/* packages

Comment (chenjiahan) [reply tests/integration/storybook/package.json:33]: It is expected that only `@modern-js/storybook` is installed and everything is out of the box.

File: packages/storybook/builder/package.json (hunk 2 above)
Comment (JSerFeng) [reply packages/storybook/builder/package.json:35]: Yes
image

File: tests/integration/storybook/package.json (hunk 4 above)
Comment (JSerFeng) [reply tests/integration/storybook/package.json:33]: Let me try

File: packages/storybook/builder/src/plugin-storybook.ts (hunk 3 above)
Comment (JSerFeng) [reply packages/storybook/builder/src/plugin-storybook.ts:246]: It's easy to support, so I added it too

File: packages/builder/builder-rspack-provider/src/plugins/swc.ts [hunk 5]
```diff
@@ -19,7 +19,7 @@ import { Builtins } from '@rspack/core';
  * Provide some swc configs of rspack
  */
 export const builderPluginSwc = (): BuilderPlugin => ({
-  name: 'builder-plugin-swc',
+  name: 'builder-plugin-swc-rspack',
```
[the remark this answers is not in the document]
Comment (JSerFeng) [reply packages/builder/builder-rspack-provider/src/plugins/swc.ts:22]: I think it is builder-plugin of swc of rspack

File: packages/runtime/plugin-runtime/package.json [hunk 6]
```diff
@@ -27,6 +27,7 @@
     ".": {
       "types": "./dist/types/index.d.ts",
       "jsnext:source": "./src/index.ts",
+      "node": "./dist/cjs/index.js",
```
[the remark this answers is not in the document]
Comment (JSerFeng) [reply packages/runtime/plugin-runtime/package.json]: Done

File: packages/storybook/builder/src/plugin-storybook.ts (hunk 3 above)
Comment (JSerFeng) [reply packages/storybook/builder/src/plugin-storybook.ts:288]: Update: we better do so, users can still stay old version stories mdx

File: packages/builder/builder-webpack-provider/src/types/thirdParty/index.ts [hunk 7]
```diff
@@ -13,7 +17,13 @@ export type { BabelConfigUtils } from '@modern-js/babel-preset-app';
 
 export type TSLoaderOptions = Partial<RawTSLoaderOptions>;
 
-export type { webpack, WebpackChain, WebpackConfig };
+export type {
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/types/thirdParty/index.ts:20]: we can import from `@modern-js/builder-webpack-provider/webpack`

**Comment by JSerFeng** [comment]
The legacy storybook plugin is moved to another repo https://github.com/web-infra-dev/legacy-storybook

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**